engine/flink/components/dev-table/src/main/scala/pl/touk/nussknacker/engine/flink/table/extractor/DataSourceSqlExtractor.scala

```diff
@@ -0,0 +1,127 @@
+
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.flink.table.api.{EnvironmentSettings, TableEnvironment, TableResult}
+import org.apache.flink.table.catalog.{CatalogBaseTable, ObjectPath}
+import org.apache.flink.table.types.DataType
+import pl.touk.nussknacker.engine.api.typed.typing.TypingResult
+import pl.touk.nussknacker.engine.flink.table.extractor.SqlDataSourceConfig.Connector
+import pl.touk.nussknacker.engine.flink.table.extractor.SqlStatementReader.SqlStatement
+import pl.touk.nussknacker.engine.flink.table.extractor.TypeExtractor.extractTypedSchema
+
+import scala.util.Try
+
+// TODO: additional validation (maybe outside scope of this class' responsibility):
+//  - duplicate table names
+//  - custom validation per connector
+object DataSourceSqlExtractor extends LazyLogging {
+
+  import scala.jdk.CollectionConverters._
+  import scala.jdk.OptionConverters.RichOptional
+
+  private val connectorKey        = "connector"
+  private val builtInCatalogName  = "defaultCatalog"
+  private val buildInDatabaseName = "defaultDatabase"
+
+  def extractTablesFromFlinkRuntime(
+      createTableStatements: List[SqlStatement]
+  ): List[Either[SqlExtractorError, SqlDataSourceConfig]] = {
+    val settings = EnvironmentSettings
+      .newInstance()
+      .withBuiltInCatalogName(builtInCatalogName)
+      .withBuiltInDatabaseName(buildInDatabaseName)
+      .build()
+    val tableEnv = TableEnvironment.create(settings)
+
+    /**
+     * Access to `CatalogBaseTable` (carrying schema details such as connector, format, and options) requires the name
+     * of the catalog/database. This code uses the default catalog and database.
+     * Assumptions:
+     *  - The executed sql statements create tables in the default catalog and database
+     *
+     * @see <a href="https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/table/catalogs/#registering-a-catalog">Flink docs source </a>
+     */
+    val catalog = tableEnv.getCatalog(builtInCatalogName).toScala match {
+      case Some(value) => value
+      case None =>
+        throw new IllegalStateException(
+          "Default catalog was not found during parsing of sql for generic table components."
+        )
+    }
+
+    /**
+     *  1. Execution of arbitrary sql statement in temp context
+     *  1. Extraction of schema and metadata from the first (only) table in the env
+     *  1. Dropping the table so we have guarantee that there are is only one table in the env
+     *
+     *  Assumptions:
+     *   - There are no tables registered by default
+     **/
+    def extractOne(
+        statement: SqlStatement
+    ): Either[SqlExtractorError, SqlDataSourceConfig] = {
+      implicit val statementImplicit: SqlStatement = statement
+      for {
+        _ <- tryExecuteStatement(statement, tableEnv)
+        // Need to grab table from env because table from result has wrong schema
+        tableName <- tableEnv
+          .listTables()
+          .headOption
```

Comment:
What about the rest of the list? :)  I see that we broken a YAGNI rule a bit. IMO we should start with more naive implementation:
- execute all statements
- do the rest with all tables.

Currently, it looks like handles some edge cases but it is hard to define which one it handles and which not. We can dive into this or we can just ignore edge cases and follow the YAGNI rule

Reply:
That was a great idea, I've implemented a simpler approach with less obstructions and now theoretically even arbitrary catalogs should work (if we add dependencies to classpath).